### ISSUE TITLE
Feature request: ignore aspect ratio (was "Question: Un-uniform scale possible?")

### ISSUE THREAD

**Issue by jsongHBO** [lovell/sharp#118]
Hi folks,

How do I resize an image disrespect of aspect ratio and without cropping and embedded padding to the exact size I want?
In gm's term, it'd be resize(width, height, '!');

Thank you

**Comment by bkw** [comment]
that would be exactly what I need as well (See dupe #153)

**Comment by jcupitt** [comment]
I haven't supported this feature in vipsthumbnail because it's hard to do correctly. 

Upscaling and downscaling are rather different operations and need separate code paths. If you allow hscale and vscale to differ, you at least double the number of code paths you need (both down, both up, hscale down vscale up, vscale down, hscale up), and perhaps double again, since (in some cases) downscale will already have two paths internally. 

Supporting not-very-different hscale and vscale would be pretty easy, which would probably cover most people. I suppose it could print a warning if the scales are either side of unity or if the scales differ by more than 20%.

**Comment by bkw** [comment]
Thanks for the explaination, @jcupitt !

Unfortunately, for my (admittedly rather odd) usecase this would not be helpful. Maybe other users would still approve, dunno.

**Comment by jcupitt** [comment]
Oh, that's a shame. How large a difference do you need?

**Comment by bkw** [comment]
The thing is - I don't know! I'm mass-processing images from third party feeds and need to scale them down to a square for further processing. Since i have no control over the source images, the ratio could be anything.

If it helps anything: the target square has a size n*n with n being a power of 2, but never greater than 256:
```javascript
smallestEdge = Math.min(orig.width, orig.height, 256);
squareSize = Math.pow(2, Math.floor(Math.log(smallestEdge) / Math.log(2)));
```

**Comment by jcupitt** [comment]
Wow that is a bizarre use case :) I guess this for some sort of query-by-image-content application? If that's the case I suppose you probably wouldn't care about small errors and a simple solution would be fine.

**Comment by bkw** [comment]
You got that right :-)

I cannot really say how important "small errors" will be before I tried them out, but since the algorithm is resilient against compression artifacts and such things, it will probably be ok.

**Comment by lovell** [comment]
Very happy to accept a PR implementing this feature - there's a fairly basic [guide for contributors](https://github.com/lovell/sharp/blob/master/CONTRIBUTING.md#submit-a-pull-request-with-a-new-feature).

**Comment by blowsie** [comment]
For reference, I always used timthumb in the past, and found the `zc` (zoom and crop) options very useful and easy to understand.

http://www.binarymoon.co.uk/2011/03/timthumb-proportional-scaling-security-improvements/

**Comment by lovell** [comment]
Thanks to @skedastik this feature will be available in the forthcoming v0.10.0 release.
